### DISCUSSION TITLE
How Users can Use Plugins/Hooks

### DISCUSSION THREAD

**Discussion by JeffersGlass** [pyscript/pyscript#1785] [Q&A]
If I, as a general user, write a Plugin (i.e. not one built into this repo, but somewhere else entirely), how does a user actually use it on their page?

I see that strategy that @WebReflection and @ntoll used [their error plugin demo](https://github.com/pyscript/pyscript/pull/1677) ([HTML](https://github.com/WebReflection/pyscript/blob/df2af0ffb418610009203c2a63858a73c5160cbe/pyscript.core/test/plugin.html), [JS](https://github.com/WebReflection/pyscript/blob/df2af0ffb418610009203c2a63858a73c5160cbe/pyscript.core/test/error.js)) is essentially:

1. Use an import map to provide readable imports of `core.js` and the plugin's JS file
2. In the plugin's JS file, `import {hooks] from ...` and add to the hook collections (`onBeforeRun`, etc)
3. In a separate module script, import `core.js` and the plugin file.

Is this the recommended/best way for users to include external plugins on their pages?

(Note that I'm specifically talking about plugins that AREN'T built into the core, like our _current_ error handling is.)

**Comment by JeffersGlass** [comment]
Now just a stream-of-consciousness - for simple "in-line" hooks that do one thing, the following pattern is useful, where the hook could be [any of the PolyScript hooks](https://github.com/pyscript/polyscript/tree/main/docs#custom-scripts-options):

```html
<script type="module">
  import { config, hooks } from "https://pyscript.net/snapshots/2023.09.1.RC2/core.js"
  hooks.onInterpreterReady.add((wrap, element) => {
    console.log({wrap}); 
    console.log({element})
  })
</script>
```
  - `wrap` contains references to interpreter, io, XWorker, run, runAsync, etc
  - `element` is the target of the currently executing script tags

~Note that this *replaces* the normal `<script type="module" src="...">` tag for loading pyscript onto the page.~

**Reply by WebReflection** [reply]
there's nothing wrong in this example ... was this a question or anything else you're concerned about? in ESM you can load the same module dozen times it's always the same anyway so this doesn't replace `<script type="module" src="https://pyscript.net/snapshots/2023.09.1.RC2/core.js"></script>` it can be added before or after that script regardless.

The hint around import maps is that anyone can then `<script type="module" src="pyscript">` or `import * from 'pyscript'` and everyone will be assured the module is always the same.

Not sure I've helped here or created more confusion, but that's how modules work on the Web.

**Reply by JeffersGlass** [reply]
It helps a lot! To be honest, as someone who is far more 'Python' than 'Web' (and that seems to true of a lot of our users), I wasn't entirely confident that there wasn't some trick about imports, or order of execution, or hooks, or some other thing I wasn't thinking about that would break this. Appreciate the confirmation!

**Comment by ntoll** [comment]
@JeffersGlass documenting a plugin "howto" is next on my todo list and would love your input..! Could you, me and perhaps @WebReflection and @fpliger catch up over Zoom to "get our story straight"..? ;-)

**Reply by JeffersGlass** [reply]
I would love that!

**Reply by fpliger** [reply]
Let's do it! Pleeeeeease! :)

**Comment by WebReflection** [comment]
There are 2 blockers for this story to be "*right*" by any defintion:

  * [renaming hooks](https://github.com/pyscript/polyscript/issues/50)
  * [make plugins work in workers too](https://github.com/pyscript/pyscript/issues/1678)

The first one breaks all docs in all places but imho is the right thing to do and sooner than later.

The second one is extremely complex to solve or provide in a super user-friendly way, because everything is defined on the main thread, but callbacks and plugins code trusting `document` or even `window` can't be written or land, as they are, in a worker + worker hooks **MUST** be strings but we cannot evaluate *JS* (or better, we really shouldn't) and callbacks cannot trespass Main / Worker boundaries ... I would lie if I'd tell you I have a solution for that, the truth is that I am still wrapping my head around something *reasonable* that is not extremely hard to either use or document as offered API which will completely replace the old API around plugins, as workers never existed there so life was definitively easier from all sides (plus multiple interpreters were never a thing so things were even more easy to deal with).

To me the plugins for 3rd party (because internal plugins have the privilege to deal directly with the code as module) should be something like this:

  * it should be possible to specify the *custom type* target: i.e. `py` and `mpy` or both or anything else in the future (implicit when `py-config` or `mpy-config` are around though ... so maybe this is solved already)
  * it should be as seamless as possible to guarantee references work on both main and workers (so they need to import `window` and `document`, as example, from our own module somehow ... as we do in Python)
  * it should be possible to define either JS or Python code or even both, for more convoluted / advanced use cases
  * it should be clear that the JS code can run only on main
  * we might find a way to instead offer an abstraction that takes care of orchestrating callbacks behind the `sync` magic we offer already but it's a whole new project that could be a standalone one to me
  * last, but not least, a plugin should be able to describe itself as *main* only, *worker* only, or both ... with its own internal orchestration

Please note I am thinking as I am writing or as we speak how to fulfill all these points **but** if you have questions, concerns, or literally any idea how to do this **technically speaking** I am all ears for you.

This is also to suggest whatever video call we do it's better next week than tomorrow, as I won't have any answer tomorrow (or maybe I will ... but it's a short notice) while I really want to have all answers next time we talk about it and I also want to have all hooks in a better shape so it's easier to understand any architecture around these too.

**Reply by ntoll** [reply]
@WebReflection I hear you!

I've scheduled a meeting Friday Euro PM (check your email). The important thing is, given the constraints you mention, we get our story straight and do "the right thing" :tm: from the start, and refine/adjust things *before* we go tell everyone how to write plugins! :wink: This is why getting our story straight is so important.

Some thoughts on what you write:

* Renaming hooks - I say we agree on the names on Friday (!) and make it so.
* Web workers - in the first instance, and to keep things well bounded and simple, we just say that plugins only ever work on the main thread (this is no different to how classic PyScript worked). I guess workers _could_ spawn new web workers if that's something they need, but that's on the head of the plugin writer. Also, we can't be 100% certain we ought to be able to support workers... if folks want workers they'll tell us, and when they do, we can engage with that process. Otherwise, let's KISS (keep is simple, silly!). :+1:
* Regarding you bullet point list of the different aspects of a plugin - I'm broadly in agreement and they make sense to me... like I said, we need to get our story straight and my intuition is to start simple/small and refine. Let's do that on Friday.

**Reply by WebReflection** [reply]
the outer points are OK to me but the inner one is a bit "*naive*" due current work on `py-terminal` which is the plugin that more than any other needs a story from a worker ... but that's not the only one, the current *error* plugin works only on main but that's because code is orchestrated to reach it from workers **although** in a worker only scenario where *main* plugins are never awaited that plugin indeed fails to show on DOM stuff and that's expected from the current logic because workers don't wait for main stuff in general.

We can flip the status-quo here and say that plugins must be awaited *before* any custom type is even defined, so that *error* would work out of the box from workers only use cases (more common in the future, imho) but that doesn't strictly solve the fact we still want to disambiguate *main* only from *workers* only plugins, as things are subtly different there.

**edit** also we need to consider how 3rd party implements their own plugins here out of `py-config` or `mpy-config` too, otherwise it will always be too late for already bootstrapped and instrumented workers.

> I guess workers could spawn new web workers

this is not the issue we have now ... the issue is that workers are not aware at all about main plugins at this stage ... they could spawn whatever but they will still be not aware of the main plugins.

> Also, we can't be 100% certain we ought to be able to support workers

we can 100% support workers if plugins target workers ... which is my idea of disambiguation. Some plugin could easily work on workers only for many reasons, starting with the fact loading and consuming code on demand can be sync in worker, but it can never be sync in the main, as that would be blocking.

> if folks want workers they'll tell us

@antocuni already told me many times he wants stuff to work in workers "first" so we're dog-fooding already users' expectations.

> Otherwise, let's KISS (keep is simple, silly!).

That's the **current status** and the fact we nede a call suggests to me it's not simple already ... although @JeffersGlass got it all right so I am still not sure, if we'd like to keep it simple, what is it that we need to discuss, as the simple approach is already demoed, understood, and OK(ish?) enough for the time being. *workers* though are a different beast and we should not underestimate caveats around these for the plugin story.

I hope my answer helps further getting the story right.

**Reply by JeffersGlass** [reply]
To be honest, I had lost track of all of the complications around hooks in workers... thanks for linking those threads back in @WebReflection.

But also, I think both @ntoll and I are saying that we *don't* feel that we understand the current simple approach well enough to explain it, or maybe we're feeling that there could be gaps in our knowledge. And that we'd love some time with you to make sure we really *do* understand, so we can make sure that what we're putting into the docs/examples/Discord etc. is accurate to what we have now. 

If we focused on that on Friday, I'd be more than happy. And you're right, we should 100% keep thinking about the plugins+workers story. But for me, I would love to make sure I fully understand the current approach, so I can go off and try to build a bunch of things/help others build things, which I think will help me have a stronger/clearer view of what we need out of plugins+workers.

**Reply by fpliger** [reply]
This is a great discussion and due to the complexity of the work, it's complicated to have a discussion here. I agree that starting to unravel the issue on a call on Friday is a great first step (in fact, there's a lot to unpack above).

Also think that, as a principle, the best idea is to start supporting the simplest scenario (in a way that it doesn't lock ourself in a place that we can't support more in the future) while we learn more on how we can support the complex cases (plugins in workers) as we work in future releases. My hope is that as we work on plugins that need to extend to workers (per @WebReflection comment, `terminal` and `errors` plugins) we can learn more and possibly find ways.

**Reply by WebReflection** [reply]
Yesterday I had a look and thought "I just wait for plugins before defining the custom type" then again reality kicked in because plugins do in fact use dependencies from the main file, which is OK as these are not really circular (they are imported asynchronously only if desired) but I also need to somehow disassociate this relation or we can't basically preload plugins ... I have ideas though and I'll implement these *before* tomorrow call.

**Reply by ntoll** [reply]
Bravo @WebReflection - as @fpliger said, it's great to get movement and momentum here. I also think we're all mostly on the same page when it comes to approach (start simple) and features (empowering folks to make full use of the platform), but we need to get our story straight so we all know what we're doing. :wink:

**Reply by WebReflection** [reply]
FYI the MR that ensures plugins come before custom elements are bootstrapped is here: https://github.com/pyscript/pyscript/pull/1788
